src/EventSubscriber/WebSSOEventSubscriber.php

```
@@ -0,0 +1,312 @@
+            if (!empty($sessionExpireOption)) {
+                $this->sessionExpirationDelay = (int) $sessionExpireOption[0]->getValue();
+            }
+            $token = new ProviderToken(
+                $webSSOConfigurationId,
+                $sessionId,
+                new \DateTime(),
+                (new \DateTime())->add(new \DateInterval('PT' . $this->sessionExpirationDelay . 'M'))
+            );
+            $this->createAuthenticationTokens(
+                $sessionId,
+                $user,
+                $token,
+                null,
+                $clientIp,
+            );
+        }
+    }
+
+    /**
+     * create Authentication tokens
+     *
+     * @param string $sessionToken
+     * @param ContactInterface $contact
+     * @param ProviderToken $providerToken
+     * @param ProviderToken|null $providerRefreshToken
+     * @param string|null $clientIp
+     * @throws AuthenticationException
+     */
+    private function createAuthenticationTokens(
+        string $sessionToken,
+        ContactInterface $contact,
+        ProviderToken $providerToken,
+        ?ProviderToken $providerRefreshToken,
+        ?string $clientIp,
+    ): void {
+        $isAlreadyInTransaction = $this->dataStorageEngine->isAlreadyinTransaction();
+
+        if (!$isAlreadyInTransaction) {
+            $this->dataStorageEngine->startTransaction();
+        }
+        try {
+            $session = new Session($sessionToken, $contact->getId(), $clientIp);
+            $this->sessionRepository->addSession($session);
+            $this->authenticationRepository->addAuthenticationTokens(
+                $sessionToken,
+                $providerToken->getId(),
+                $contact->getId(),
+                $providerToken,
+                $providerRefreshToken
+            );
+            if (!$isAlreadyInTransaction) {
+                $this->dataStorageEngine->commitTransaction();
+            }
+        } catch (\Exception) {
```

Exception is lost, you should at least log it

---

good point, i forgot to add log in all the flow